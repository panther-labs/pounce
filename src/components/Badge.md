```jsx harmony
<div>
  <Badge variant="info" mb={2} />
  <Badge variant="error-low" mb={2} />
  <Badge variant="error-medium" mb={2} />
  <Badge variant="error-high" mb={2} />
  <Badge variant="error-critical" mb={2} />
</div>
```
