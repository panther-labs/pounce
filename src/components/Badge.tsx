import React from 'react';
import Text from './Text';

interface BadgeProps {
  /** The style of the badge */
  variant: 'info' | 'error-low' | 'error-medium' | 'error-high' | 'error-critical';
}

/**
 * A badge is simply a visual label to associate with something else
 */
const Badge: React.FC<BadgeProps> = ({ variant, ...rest }) => {
  const variantProps = (function() {
    switch (variant) {
      case 'info':
        return {
          border: '1px solid',
          borderColor: 'grey200',
          bg: 'transparent',
          color: 'grey300',
          children: 'INFO',
        };
      case 'error-low':
        return { bg: 'grey100', color: 'grey300', children: 'LOW' };
      case 'error-medium':
        return { bg: 'blue100', color: 'blue300', children: 'MEDIUM' };
      case 'error-high':
        return { bg: 'red100', color: 'red300', children: 'HIGH' };
      case 'error-critical':
      default:
        return { bg: 'red300', color: 'white', children: 'CRITICAL' };
    }
  })();

  return (
    <Text
      width={62}
      textAlign="center"
      fontSize={0}
      fontWeight="bold"
      borderRadius="small"
      lineHeight="12px"
      px={3}
      py={1}
      {...variantProps}
      {...rest}
    />
  );
};

export default Badge;
