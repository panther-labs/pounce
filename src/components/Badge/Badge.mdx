A Badge should be used to render statuses

A badge can have different colors:

```typescript jsx
<Flex spacing={2}>
  <Badge color="white">Primary</Badge>
  <Badge color="red-300">Red</Badge>
  <Badge color="blue-400">Blue</Badge>
</Flex>
```

A badge can have different sizes:

```typescript jsx
<Flex align="center" spacing={2}>
  <Badge color="blue-400" size="small">
    Small
  </Badge>
  <Badge color="blue-400" size="medium">
    Medium
  </Badge>
</Flex>
```

A badge can be `emphasized`:

```typescript jsx
<SimpleGrid spacing={3} columns={3}>
  <Badge stretch emphasized color="red-300">
    INFO
  </Badge>
  <Badge stretch emphasized color="teal-500">
    LOW
  </Badge>
  <Badge stretch emphasized color="blue-400">
    WHATEVER
  </Badge>
</SimpleGrid>
```
