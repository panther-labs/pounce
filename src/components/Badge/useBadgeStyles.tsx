import { BadgeProps } from './Badge';
import { FlexProps } from 'components/Flex';

type UseBadgeStylesProps = Pick<BadgeProps, 'variant' | 'color'>;
type UseBadgeStylesPayload = Partial<FlexProps>;

const useBadgeStyles = ({ variant, color }: UseBadgeStylesProps): UseBadgeStylesPayload => {
  switch (variant) {
    case 'outline':
      return { border: '1px solid', borderColor: color, bg: 'transparent' as const };
    case 'solid':
    default:
      return { bg: color };
  }
};

export default useBadgeStyles;
