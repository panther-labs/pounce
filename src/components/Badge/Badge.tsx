import React from 'react';
import { Theme } from 'theme';
import useBadgeStyles from './useBadgeStyles';
import Box from 'components/Box';

export interface BadgeProps {
  /** The color theme of the badge */
  color: keyof Theme['colors'];

  /** The style of the badge */
  variant?: 'solid' | 'outline';

  /** @ignore */
  children: React.ReactNode;
}

/** A badge is simply a visual label to accompany & characterize a certain text*/
const Badge = React.forwardRef<HTMLElement, BadgeProps>(function Badge(
  { color, variant = 'solid', children, ...rest },
  ref
) {
  const variantStyles = useBadgeStyles({ variant, color });

  return (
    <Box
      display="flex"
      role="status"
      aria-atomic="true"
      ref={ref}
      cursor="default"
      width="fit-content"
      minWidth="85px"
      textAlign="center"
      fontWeight="bold"
      borderRadius="pill"
      alignItems="center"
      justifyContent="center"
      fontSize="small"
      py={1}
      px={4}
      {...variantStyles}
      {...rest}
    >
      {children}
    </Box>
  );
});

export default React.memo(Badge);
