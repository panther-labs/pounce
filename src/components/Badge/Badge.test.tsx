import React from 'react';
import { renderWithTheme, act, waitFor, AxeResults } from 'test-utils';
import { axe } from 'jest-axe';

import Badge from './Badge';

it('renders', () => {
  const { container } = renderWithTheme(
    <>
      <Badge color="red-300">INFO</Badge>
      <Badge color="teal-500">LOW</Badge>
      <Badge color="blue-400">WHATEVER</Badge>
    </>
  );
  expect(container).toMatchSnapshot();
});

it('aligns with a11y', async () => {
  const { container } = renderWithTheme(
    <>
      <Badge color="red-300">INFO</Badge>
      <Badge color="teal-500">LOW</Badge>
      <Badge color="blue-400">WHATEVER</Badge>
    </>
  );
  let results: AxeResults;
  await act(async () => {
    results = await axe(container);
  });
  await waitFor(() => {
    expect(results).toHaveNoViolations();
  });
});
