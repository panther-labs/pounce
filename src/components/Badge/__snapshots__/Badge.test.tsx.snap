// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.pounce-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: default;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  text-align: center;
  font-weight: 400;
  text-transform: none;
  border: 1px solid;
  border-radius: 2px;
  border-color: #d64242;
  background-color: rgba(214,66,66,0.3);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  font-size: 0.75rem;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 6px;
  padding-right: 6px;
}

.pounce-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: default;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  text-align: center;
  font-weight: 400;
  text-transform: none;
  border: 1px solid;
  border-radius: 2px;
  border-color: #09a084;
  background-color: rgba(9,160,132,0.3);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  font-size: 0.75rem;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 6px;
  padding-right: 6px;
}

.pounce-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: default;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  text-align: center;
  font-weight: 400;
  text-transform: none;
  border: 1px solid;
  border-radius: 2px;
  border-color: #0081c5;
  background-color: rgba(0,129,197,0.3);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  font-size: 0.75rem;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 6px;
  padding-right: 6px;
}

<div>
  <div
    aria-atomic="true"
    class="pounce-0"
    role="status"
  >
    INFO
  </div>
  <div
    aria-atomic="true"
    class="pounce-1"
    role="status"
  >
    LOW
  </div>
  <div
    aria-atomic="true"
    class="pounce-2"
    role="status"
  >
    WHATEVER
  </div>
</div>
`;
